# Objective
**Build and compare Scikit-learn preprocessing pipelines and evaluate two classification models.**

# Part A - Data Loading and Preprocessing with Scikit-learn

**Task 1 - Load and Understand the Dataset**
* Load hotel_bookings.csv. Use head(), shape, info(), describe(), and dtypes.
* Check the class distribution of is_canceled and use it as the target y.
* Create X from the remaining usable features. Identify numerical and categorical columns.

In [ ]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv('hotel_bookings.csv')
print(data.head(5))
print(data.shape)
print(data.info())
print(data.describe())
print(data.dtypes)

          hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1     

In [4]:
# class distribution of is_cancelled
print("\nClass Distribution:")
print(data["is_canceled"].value_counts())


Class Distribution:
is_canceled
0    75166
1    44224
Name: count, dtype: int64


In [5]:
# Target variable
y = data["is_canceled"]

# Remaining columns are input features
X = data.drop(columns=["is_canceled"])

In [6]:
# Identify numerical and categorical columns
numerical_columns = X.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

categorical_columns = X.select_dtypes(
    include=["object"]
).columns.tolist()

print("\nNumerical Columns:")
print(numerical_columns)

print("\nCategorical Columns:")
print(categorical_columns)


Numerical Columns:
['lead_time', 'arrival_date_year', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'agent', 'company', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']

Categorical Columns:
['hotel', 'arrival_date_month', 'meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'customer_type', 'reservation_status', 'reservation_status_date']


C:\Users\devan\AppData\Local\Temp\ipykernel_8856\2736801362.py:6: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = X.select_dtypes(


**Task 2 - Missing Values, Leakage, and Outliers**
* Find the missing-value count and percentage for every column.
* Identify columns with very high missingness. Drop them only if justified and state the reason.
* Remove columns that directly reveal the final booking outcome, such as reservation_status and
reservation_status_date.
* Check selected numerical features for outliers using boxplots and/or the IQR method.
* Remove only clear/extreme outliers and report how many rows were removed.

In [7]:
# Missing value count
missing_count = data.isnull().sum()

print("\nMissing Values:")
print(missing_count)

# Missing value percentage
missing_percentage = (data.isnull().sum() / len(data)) * 100

print("\nMissing Value Percentage:")
print(missing_percentage)


Missing Values:
hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340

- children has 4 missing values, it is not very high so we don't need to remove it
- country has 488 missing values, it is higher than children but decision has to be made that we should remove it not based on its influence on other columns
- agent column has 16340 missing value, it is very much higher we need to remove this column because, it will not contribute anyway in model and only increase the computation. so we remove this column
- similar for company

In [8]:
df = data.drop(columns=["company", "agent"])